packages/http/fetch/src/middlewares/authorizationHandler.ts

```diff
@@ -0,0 +1,139 @@
+ * -------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.
+ * See License in the project root for license information.
+ * -------------------------------------------------------------------------------------------
+ */
+
+import { type RequestOption } from "@microsoft/kiota-abstractions";
+import { Span, trace } from "@opentelemetry/api";
+
+import { getObservabilityOptionsFromRequest } from "../observabilityOptions";
+import type { Middleware } from "./middleware";
+import type { FetchRequestInit } from "../utils/fetchDefinitions";
+import { getRequestHeader, setRequestHeader } from "../utils/headersUtil";
+import { BaseBearerTokenAuthenticationProvider, Headers, RequestInformation } from "@microsoft/kiota-abstractions/src";
+
+export class AuthorizationHandler implements Middleware {
+	next: Middleware | undefined;
+
+	/**
+	 * A member holding the name of content range header
+	 */
+	private static readonly AUTHORIZATION_HEADER = "Authorization";
+
+	public constructor(private readonly authenticationProvider: BaseBearerTokenAuthenticationProvider) {
+		if (!authenticationProvider) {
+			throw new Error("authenticationProvider cannot be undefined");
+		}
+	}
+
+	public execute(url: string, requestInit: RequestInit, requestOptions?: Record<string, RequestOption>): Promise<Response> {
+		const obsOptions = getObservabilityOptionsFromRequest(requestOptions);
+		if (obsOptions) {
+			return trace.getTracer(obsOptions.getTracerInstrumentationName()).startActiveSpan("authorizationHandler - execute", (span) => {
+				try {
+					span.setAttribute("com.microsoft.kiota.handler.authorization.enable", true);
+					return this.executeInternal(url, requestInit as FetchRequestInit, requestOptions, span);
+				} finally {
+					span.end();
+				}
+			});
+		}
+		return this.executeInternal(url, requestInit as FetchRequestInit, requestOptions, undefined);
+	}
+
+	private async executeInternal(url: string, fetchRequestInit: FetchRequestInit, requestOptions?: Record<string, RequestOption>, span?: Span): Promise<Response> {
+		if (this.authorizationIsPresent(fetchRequestInit)) {
+			span?.setAttribute("com.microsoft.kiota.handler.authorization.token_present", true);
+			return await this.next!.execute(url, fetchRequestInit as RequestInit, requestOptions);
+		}
+
+		const requestInformation = new RequestInformation();
+		requestInformation.URL = url;
+		requestInformation.headers = new Headers();
+		if (fetchRequestInit.headers !== undefined) {
+			const headers = fetchRequestInit.headers as Record<string, string>;
+			for (const key of Object.keys(headers)) {
+				const value = headers[key];
+				requestInformation.headers.add(key, value);
+			}
+		}
+
+		// add all the headers to fetchRequestInit
+		if (!fetchRequestInit.headers) {
+			fetchRequestInit.headers = {};
+		}
+
+		await this.authenticateRequest(requestInformation);
```

Comment:
we should instead add a getter for the access token provider on the authentication provider and just pass the URI

Reply:
had to change the constructor to use `AccessTokenProvider` directly

Reply:
sorry if I wasn't clear, the constructor should still accept a base bearer auth provider because this is what the user will have at hand. But that class should expose a property to let us read the access token provider so we can use it here.
Does this make things a bit more clear?

Reply:
makes sense

Reply:
while I'm at it, we'll need an optional parameter for a base bearer auth provider in the kiota client factory, and for this middleware to be part of the default factory.

Reply:
Made both changes